### PR TITLE
feat: cache etag support

### LINKED ISSUE #3718: Easier way to attach `ETag` to `If-None-Match` after #3562
## This would solve...

#3562 introduces HTTP Caching (RFC9110) to undici, which also implements HTTP 304 with `If-Modified-Since`. But it seems that `ETag` is missing.

## The implementation should look like...

IMHO #3562 should not be delayed anymore, this could be implemented after #3562 is merged.

The `undici` could check the stored `ETag` header from the cache store and use that to negotiate with the `If-None-Match` in future requests. If the server returns HTTP 304, undici would then return the cached (transparent) HTTP 200 response to the client (the same behavior of `fetch` in browsers).

## Additional context

#3562 

https://github.com/JakeChampion/fetch/issues/241

`ETag` is described in RFC9110 here: https://httpwg.org/specs/rfc9110.html#field.etag
`If-None-Match` is described in RFC9110 here: https://httpwg.org/specs/rfc9110.html#field.if-none-match
HTTP 304 is described in RFC9110 here: https://httpwg.org/specs/rfc9110.html#status.304
The handling of HTTP 304 is described in RFC9110 here: https://httpwg.org/specs/rfc9111.html#freshening.responses

### PR DESCRIPTION
**Pull request by flakey5**
Re http caching, adds support for etags for when we need to revalidate a cached response.

No breaking changes to the public api. the `etag` property is nullable so a cache store _could_ just not implement it and it will be fine. However, I did add it to the tests as if it was required just so we can ensure the builtin cache stores fully support the behavior.

This only checks if the etag exists in the response's headers and ignores it if it's in the trailers for simplicity sakes (as allowed by the spec in [8.8.3](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.8.3-8))

Closes #3718

### CODE REVIEW

File: lib/handler/cache-handler.js [hunk 1]
```diff
@@ -126,15 +126,25 @@ class CacheHandler extends DecoratorHandler {
...
         vary: varyDirectives,
         cachedAt: now,
         staleAt,
         deleteAt
-      })
+      }
+
+      // Check for the response's etag header, ignore if there's multiple
+      if (typeof headers.etag === 'string') {
+        value.etag = headers.etag
```
Comment (KhafraDev) [inline lib/handler/cache-handler.js:144]: Shouldn't etag be validated here? https://httpwg.org/specs/rfc9110.html#field.etag

Comment (metcoder95) [reply lib/handler/cache-handler.js:144]: and also account for weak etags

**Comment by mcollina** [comment]
I think this is a bit too light on tests, and I'm not sure if this implementation is correct.

Can you add some tests where the response is a 304 thanks to if-none-match?

**Comment by SukkaW** [comment]
> I think this is a bit too light on tests, and I'm not sure if this implementation is correct.
> 
> Can you add some tests where the response is a 304 thanks to if-none-match?

Also, we could test how undici responds to different server implementations. I myself have run into situations when locally stored strong ETag matches the server-sent one yet the server is returning 200 with the same ETag, instead of 304.

**Review by ronag — APPROVED** [review]

File: lib/util/cache.js [hunk 2]
```diff
@@ -201,6 +201,48 @@ function parseVaryHeader (varyHeader, headers) {
...
+
+  if (etag[0] === '"' && etag[etag.length - 1] === '"') {
+    if (etag[1] === '"' || etag.startsWith('"W/')) {
+      // ETag: ""asd123"" or ETag: "W/"asd123"", kinda undefined behavior in the
+      //  spec. Some servers will accept these while others don't.
+      return false
+    }
+
+    // ETag: "asd123"
+    return true
```
Comment (metcoder95) [inline lib/util/cache.js:228]: `nit`:
```suggestion
    // ETag: ""asd123"" or ETag: "W/"asd123"", kinda undefined behavior in the
    //  spec. Some servers will accept these while others don't.
    // ETag: "asd123"
    return !(etag[1] === '"' || etag.startsWith('"W/'))
```

File: lib/util/cache.js [hunk 3]
```diff
@@ -201,6 +201,48 @@ function parseVaryHeader (varyHeader, headers) {
...
+  }
+
+  if (etag.startsWith('W/"') && etag[etag.length - 1] === '"') {
```
Comment (metcoder95) [inline lib/util/cache.js:231]: `nit`: Similar suggestion as per my previous comment

**Review by metcoder95 — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
@metcoder95 I'm landing this as it has been around for a while.

**Comment by mcollina** [comment]
@flakey5 this now conflicts

**Review by mcollina — APPROVED** [review]
lgtm
